discovery-provider/src/tasks/index.py

```diff
@@ -365,307 +499,106 @@ def index_blocks(self, db, blocks_list):
-
-            # Without this check we may end up duplicating an insert operation
-            block_model = Block(
-                blockhash=web3.toHex(block.hash),
-                parenthash=web3.toHex(block.parentHash),
-                number=block.number,
-                is_current=True,
-            )
-
-            # Update blocks table after
-            assert (
-                current_block_query.count() == 1
-            ), "Expected single row marked as current"
-
-            previous_block = current_block_query.first()
-            previous_block.is_current = False
-            session.add(block_model)
-
-            user_factory_txs = []
-            track_factory_txs = []
-            social_feature_factory_txs = []
-            playlist_factory_txs = []
-            user_library_factory_txs = []
-            user_replica_set_manager_txs = []
-
-            tx_receipt_dict = fetch_tx_receipts(self, block.transactions)
-
-            # Sort transactions by hash
-            sorted_txs = sorted(block.transactions, key=lambda entry: entry["hash"])
 
+        with db.scoped_session() as session, challenge_bus.use_scoped_dispatch_queue():
             skip_tx_hash = save_and_get_skip_tx_hash(session, redis)
-            # Parse tx events in each block
-            for tx in sorted_txs:
-                tx_hash = web3.toHex(tx["hash"])
-                tx_target_contract_address = tx["to"]
-                tx_receipt = tx_receipt_dict[tx_hash]
-
-                # Skip in case a transaction targets zero address
-                if tx_target_contract_address == zero_address:
-                    logger.info(
-                        f"index.py | Skipping tx {tx_hash} targeting {tx_target_contract_address}"
-                    )
-                    continue
-
-                if skip_tx_hash == "commit":
-                    # The whole block is worth skipping because we failed at the database commit
-                    logger.info(f"index.py | Skipping all txs in block {block.hash}")
-                    break
-
-                if skip_tx_hash is not None and skip_tx_hash == tx_hash:
-                    logger.info(f"index.py | Skipping tx {tx_hash}")
-                    continue
-
-                # Handle user operations
-                if (
-                    tx_target_contract_address
-                    == get_contract_addresses()["user_factory"]
-                ):
-                    logger.info(
-                        f"index.py | UserFactory contract addr: {tx_target_contract_address}"
-                        f" tx from block - {tx}, receipt - {tx_receipt}, adding to user_factory_txs to process in bulk"
+            skip_whole_block = skip_tx_hash == "commit"  # db tx failed at commit level
+            if skip_whole_block:
+                logger.info(f"index.py | Skipping all txs in block {block.hash}")
+            else:
+                add_indexed_block_to_db(session, block)
```

Comment:
can this be moved down later in the function after the async work so it's near where we process the other state changes / touch the db? i think this being so much earlier in the fn was what led us to breaking this out into two sessions in the first place

Reply:
Sure! I can move this to right before process state changes as that part has logic that depends on it. In any case, the session needs to be opened early on to handle block/tx skips and can't close until the end, so the DB block tx will be rolled back if any issues occur during downstream processing.